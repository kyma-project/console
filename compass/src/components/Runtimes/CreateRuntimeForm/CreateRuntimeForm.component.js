import React, { useRef, useEffect, useState } from 'react';
import './CreateRuntimeForm.scss';
import { InlineHelp } from 'fundamental-react/lib/InlineHelp';

const CreateRuntimeForm = ({
  formElementRef,
  onChange,
  onCompleted,
  onError,
  addRuntime,
  getRuntimeNames,
}) => {
  const formValues = {
    name: useRef(null),
    description: useRef(null),
  };

  const [runtimeNames, setRuntimeNames] = useState([]);

  useEffect(() => {
    if (Array.isArray(runtimeNames) && !runtimeNames.length)
      fetchRuntimeNames();
  });

  const fetchRuntimeNames = async () => {
    const allRuntimes = await getRuntimeNames.refetch();
    if (
      !allRuntimes.errors &&
      allRuntimes &&
      allRuntimes.data &&
      allRuntimes.data.runtimes &&
      allRuntimes.data.runtimes.data
    ) {
      setRuntimeNames(allRuntimes.data.runtimes.data.map(r => r.name));
    } else {
      setRuntimeNames(null);
    }
  };

  const handleFormSubmit = async e => {
    e.preventDefault();
    const runtimeName = formValues.name.current.value;
    try {
      await addRuntime({
        name: runtimeName,
        description: formValues.description.current.value,
        labels: { test: ['hello', 'there'] },
      });
      onCompleted(runtimeName, `Runtime created succesfully`);
    } catch {
      onError(runtimeName, `The runtime could not be created succesfully`);
    }
  };

  const handleNameChanged = async () => {
    formValues.name.current.setCustomValidity(
      runtimeNames.includes(formValues.name.current.value)
        ? 'This name is already taken by another runtime'
        : '',
    );
  };

  const nameField = () => (
    <>
      <label className="fd-form__label" htmlFor="runtime-name">
        Name:
        <InlineHelp
          placement="bottom-right"
          text="Name must be no longer than 63 characters, must start and end with a lowercase letter or number, and may contain lowercase letters, numbers, and hyphens."
        />
      </label>
      <input
        className="fd-form__control"
        ref={formValues.name}
        type="text"
        id="runtime-name"
        placeholder="Runtime name"
        aria-required="true"
        required
        pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
        onChange={e => handleNameChanged(e.target)}
      />
    </>
  );

  return (
    <form
      onChange={onChange}
      ref={formElementRef}
      style={{ width: '30em' }}
      onSubmit={handleFormSubmit}
    >
      <div className="fd-form__set">
        <div className="fd-form__item">{nameField()}</div>
        <div className="fd-form__item">
          <label className="fd-form__label" htmlFor="runtime-desc">
            Description:
            <InlineHelp placement="bottom-right" text="Must not be empty" />
          </label>

          <input
            className="fd-form__control"
            ref={formValues.description}
            type="text"
            id="runtime-desc"
            placeholder="Runtime description"
            aria-required="true"
            required
          />
        </div>
      </div>
    </form>
  );
};

export default CreateRuntimeForm;
