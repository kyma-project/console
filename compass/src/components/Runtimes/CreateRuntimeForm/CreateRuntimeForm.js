import React from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';

import { Panel, PanelGrid } from 'fundamental-react/lib/Panel';

const ADD_RUNTIME = gql`
  mutation {
    createRuntime(
      in: {
        name: "testname"
        description: "Test21"
        labels: { test: ["hello", "there"] }
        annotations: { key: value }
      }
    ) {
      id
      description
      name
    }
  }
`;
let input;
const CreateRuntimeForm = () => (
  <Mutation mutation={ADD_RUNTIME}>
    {createRuntime => (
      <Panel>
        <Panel.Header>
          <Panel.Head title="Create new runtime" />
        </Panel.Header>
        <Panel.Body>
          <form
            onSubmit={e => {
              e.preventDefault();
              createRuntime({ variables: { name: 'abcd' } });
              input.value = '';
            }}
          >
            <div className="fd-form__set">
              <div className="fd-form__item">
                <label className="fd-form__label" htmlFor="input-1">
                  Name:
                </label>
                <input
                  className="fd-form__control"
                  ref={node => {
                    input = node;
                  }}
                  type="text"
                  id="input-1"
                  placeholder="Runtime name"
                  aria-required="true"
                  required
                />
              </div>
              <button className="fd-button" type="submit">
                Add Runtime
              </button>
            </div>
          </form>
        </Panel.Body>
      </Panel>
    )}
  </Mutation>
);

export default CreateRuntimeForm;
