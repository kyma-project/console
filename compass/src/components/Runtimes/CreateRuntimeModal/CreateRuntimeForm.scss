.fd-form__control:invalid {
  border: 2px solid var(--fd-color-status-3);
}
