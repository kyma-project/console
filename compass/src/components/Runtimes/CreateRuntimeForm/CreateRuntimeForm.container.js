import { graphql, compose } from 'react-apollo';

import { ADD_RUNTIME } from '../gql';

import CreateRuntimeForm from './CreateRuntimeForm.component';

export default compose(
  graphql(ADD_RUNTIME, {
    props: ({ mutate, error }) => ({
      addRuntime: data =>
        mutate({
          variables: {
            in: data,
          },
        }),
    }),
  }),
)(CreateRuntimeForm);
