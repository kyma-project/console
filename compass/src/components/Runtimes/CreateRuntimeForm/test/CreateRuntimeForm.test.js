import React from 'react';
import renderer from 'react-test-renderer';

import CreateRuntimeForm from '../CreateRuntimeForm.container';
import { MockedProvider } from 'react-apollo/test-utils';

describe('CreateRuntimeForm', () => {
  // for "Warning: componentWillReceiveProps has been renamed"
  console.warn = jest.fn();

  const emptyFn = () => {};
  const emptyRef = { current: null };

  it('Renders with minimal props', () => {
    const component = renderer.create(
      <MockedProvider addTypename={false}>
        <CreateRuntimeForm
          formElementRef={emptyRef}
          isValid={false}
          onChange={emptyFn}
          onCompleted={emptyFn}
          onError={emptyFn}
        />
      </MockedProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
  });
});
