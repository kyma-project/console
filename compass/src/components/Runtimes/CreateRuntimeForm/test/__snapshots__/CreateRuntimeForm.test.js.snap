// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRuntimeForm Renders with minimal props 1`] = `
<form
  onChange={[Function]}
  onSubmit={[Function]}
>
  <div
    className="fd-form__set"
  >
    <div
      className="fd-form__item"
    >
      <label
        className="fd-form__label"
        htmlFor="runtime-name"
      >
        Name *
        <span
          className="fd-inline-help"
        >
          <span
            className="fd-inline-help__content fd-inline-help__content--bottom-right"
          >
            Name must be no longer than 63 characters, must start and end with a lowercase letter or number, and may contain lowercase letters, numbers, and dashes.
          </span>
        </span>
      </label>
      <input
        aria-required="true"
        className="fd-form__control"
        id="runtime-name"
        pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
        placeholder="Runtime name"
        required={true}
        type="text"
      />
    </div>
    <div
      className="fd-form__item"
    >
      <label
        className="fd-form__label"
        htmlFor="runtime-desc"
      >
        Description
      </label>
      <input
        className="fd-form__control"
        id="runtime-desc"
        placeholder="Runtime description"
        type="text"
      />
    </div>
  </div>
</form>
`;

exports[`CreateRuntimeForm Renders with minimal props 2`] = `
"Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run \`npx react-codemod rename-unsafe-lifecycles\` in your project source folder.

Please update the following components: Mutation"
`;
